src/lib/emitters.rs

```diff
@@ -0,0 +1,123 @@
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file //
+// if one is included with the Software (each a “Larger Work” to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+
+use std::borrow::Cow::Owned;
+use std::fs::File;
+use std::io::{Error, Write};
+
+use dot::{Id, Edges, GraphWalk, Labeller, LabelText, Nodes, render};
+
+use ast::{Arena, EdgeId, NodeId};
+
+/// Errors produced in emitting new data.
+pub enum EmitterError {
+    /// Could not create a new file.
+    CouldNotCreateFile,
+    /// Could not write out to a file.
+    CouldNotWriteToFile,
+}
+
+/// Write out a graphviz file (in dot format) to `filepath`.
+pub fn write_dotfile_to_disk(filepath: &str,
+                             arena: Arena<String, String>)
+                             -> Result<(), EmitterError> {
+    let mut dotfile = match File::create(&filepath) {
+        Ok(f) => f,
+        Err(_) => return Err(EmitterError::CouldNotCreateFile),
+    };
+    match arena.render_dotgraph(&mut dotfile) {
+        Ok(_) => (),
+        Err(_) => {
+            return Err(EmitterError::CouldNotWriteToFile);
+        }
+    };
+    Ok(())
+}
+
+impl<'a> Labeller<'a, NodeId, EdgeId> for Arena<String, String> {
+    fn graph_id(&self) -> Id {
+        Id::new("AST").unwrap()
+    }
+
+    fn node_id(&self, id: &NodeId) -> Id {
+        // Node ids must be unique dot identifiers, be non-empty strings made up
+        // of alphanumeric or underscore characters, not beginning with a digit
+        // (i.e. the regular expression [a-zA-Z_][a-zA-Z_0-9]*).
+        Id::new(format!("N{}", id)).unwrap()
+    }
+
+    fn node_label(&self, id: &NodeId) -> LabelText {
```

Comment:
The reference is required by the trait, I think: https://docs.rs/dot/0.1.2/dot/trait.Labeller.html